### PR TITLE
[indexer] Add file granularity option

### PR DESCRIPTION
**Pull request by clee2000**
Adds a file granularity option to the indexer

Change parser to take in a granularity setting

tested in: https://github.com/pytorch/pytorch/pull/120447

### CODE REVIEW

File: indexer.py [hunk 1]
```diff
@@ -149,10 +157,17 @@ def main():
...
+        "--function", action="store_true", help="Index at function granularity"
+    )
+    group.add_argument(
+        "--file", action="store_true", help="Index at file granularity"
+    )
     args = parser.parse_args()
 
     start = time.time()
-    indexer = Indexer(args.experiment_name)
+    indexer = Indexer(args.experiment_name, file=args.file)
```
Comment (clee2000) [inline indexer.py:170]: --function doesnt do anything except in the parser, maybe i get rid of the option and just make it the default behavior?

File: test_dataset.py [hunk 2]
```diff
@@ -161,10 +171,38 @@ def __getitem__(self, idx):
...
-        tokens = self.tokenize_functions(functions)
-        function_list = list(functions.keys())
+        return self.tokenize_items(functions)
+
+
+class FileGranularityDataset(UnittestDataset):
+    def __init__(self, config):
+        super().__init__(config)
+
+    def partition_file(self, filename) -> List[FilePartition]:
```
Comment (osalpekar) [inline test_dataset.py:181]: Are we partitioning the file anywhere in the FileGranularityDataset?

File: test_dataset.py [hunk 3]
```diff
@@ -38,17 +41,9 @@ class FilePartition(NamedTuple):
 class UnittestDataset(Dataset):
```
Comment (osalpekar) [inline test_dataset.py:41]: I like the inheritance with the GraularityDatasets subclassing this one. To make things a bit more readable, do you mind not defining `UnittestDataset.__len__()` (just use `pass`) maybe. And then have the actual implementation in `FileGranularityDataset.__len__()`.

File: indexer.py [hunk 4]
```diff
@@ -9,16 +9,21 @@
...
+    FunctionGranularityDataset,
+)
 from torch.utils.data import DataLoader, DistributedSampler
 
 from transformers import AutoModelForCausalLM
 
 
 class Indexer:
-    def __init__(self, experiment_name):
+    def __init__(self, experiment_name, file: bool = False):
```
Comment (osalpekar) [inline indexer.py:23]: Maybe let's have a enum for the support granularities. And name the argument accordingly (parsing_granularity or smth). Maybe we can leverage this enum when passing the arg via argparse as well so the. arg becomes `--granularity FILE`?

**Review by osalpekar — COMMENTED** [review]
few comments, but LGTM overall

File: indexer.py (hunk 4 above)
Comment (clee2000) [reply indexer.py]: Swapped parser to take in a granularity setting but used dictionary instead of enum so I can map to the correct class

**Review by osalpekar — APPROVED** [review]
Awesome!

**Pull request merged**